Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift

```diff
@@ -130,8 +130,66 @@ open class BufferingBookmarksSynchronizer: TimestampedSingleCollectionSynchroniz
+                    return buffer.getBufferItemWithGUID(BookmarkRoots.MobileFolderGUID).bind { maybeMobileRoot -> Deferred<Maybe<Record<BookmarkBasePayload>>> in
+                        // Update (or create!) the Mobile Root folder with its new children
+                        if maybeMobileRoot.isSuccess {
+                            let mobileRoot = maybeMobileRoot.successValue!
+                            return buffer.getBufferChildrenGUIDsForParent(mobileRoot.guid) >>== { curChildren in
+                                var allChildren = curChildren
+                                let newChildren = newBookmarks.values.map { $0.guid }
+                                // We do not use a set in order to keep the previous children's order
+                                for child in newChildren {
+                                    if !allChildren.contains(child) {
+                                        allChildren.append(child)
+                                    }
+                                }
+
+                                let payload = mobileRoot.asPayloadWithChildren(allChildren)
+                                let mappedGUID = payload["id"].string ?? mobileRoot.guid
+                                return deferMaybe(Record<BookmarkBasePayload>(id: mappedGUID, payload: payload))
+                            }
+                        } else {
+                            let mobileRoot = BookmarkMirrorItem.folder(BookmarkRoots.MobileFolderGUID, modified: Date.now(), hasDupe: false, parentID: BookmarkRoots.RootGUID, parentName: "", title: "Mobile Bookmarks", description: nil, children: Array(newBookmarks.keys))
+                            return deferMaybe(Record<BookmarkBasePayload>(id: BookmarkRoots.translateOutgoingRootGUID(BookmarkRoots.MobileFolderGUID), payload: mobileRoot.asPayload()))
+                        }
+                    } >>== { mobileRootRecord in
+                        // Build a list of records to upload
+                        var records = Array(newBookmarks.values.map { bkm -> Record<BookmarkBasePayload> in
+                            let payload = bkm.asPayload()
+                            let mappedGUID = payload["id"].string ?? bkm.guid
+                            return Record<BookmarkBasePayload>(id: mappedGUID, payload: payload)
+                        })
+                        records.append(mobileRootRecord)
+
+                        let bufferOP = BufferUpdatedCompletionOp()
+                        let onUpload: (POSTResult, Timestamp?) -> DeferredTimestamp = { result, lastModified in
+                            let modified = lastModified ?? 0
+                            let uploaded = result.success.map { BookmarkRoots.translateIncomingRootGUID($0) }
+                            if (uploaded.contains(BookmarkRoots.MobileFolderGUID)) {
+                                bufferOP.mobileRootToUpdate = (mobileRootRecord.payload as MirrorItemable).toMirrorItem(modified)
```

Comment:
This betrays a weakness here.

If an upload spans several POSTs, as it will if the batching API is disabled (as it currently is), you will be leaving the server inconsistent unless all the POSTs succeed — we only upload the mobile root once, after all of its new children have been uploaded.

`uploadRecords`, which you're using here, is suitable for two kinds of use:

- Independent record synchronizers, where there is no need for inter-record consistency.
- Replayable synchronizers, where the actions of this client alone can recover from an inconsistency due to inter-record relationships.

The main bookmarks synchronizer is replayable. I'm not sure this one is; it's certainly not if you follow my suggestion earlier of avoiding collisions with remote records, because after each chunk the non-cleaned-up local records are now remote ones with which we'd collide if replayed!

My suggestion is to introduce a little simplicity here.

- Add a sibling method to `uploadRecords` that doesn't have `onUpload` and is guaranteed to make a single POST.
- Expose the `infoConfig` field for max batch size as a method on `Sync15CollectionClient`.
- Here, where you begin to upload records, ask the client how many records we can do atomically (the method above).
- Fetch that many new mobile children (minus one, obviously) from the database, and construct a single atomic payload of new children + the mobile folder.
- Upload it in one go, and move those children from the local DB to the buffer.
- Finish.

This means we'll typically upload 99 mobile bookmarks, safely, each time we sync. This has some advantages:

- Reduced chance of OOM thanks to buffering all the mobile bookmarks in memory.
- Incremental progress if mobile bookmarks gets too large.
- No chance of introducing inconsistency.
- Makes this code much more predictable and simpler: no `onUpload`, no batching, no need for multiple modified times in the buffer op.

What do you think?